packages/labs/ssr-dom-shim/src/index.ts

```diff
@@ -172,8 +172,36 @@ type CustomElementRegistration = {
   observedAttributes: string[];
 };
 
-const CustomElementRegistryShim = class CustomElementRegistry {
+type RealCustomElementRegistry = (typeof globalThis)['customElements'];
+type RealCustomElementRegistryClass =
+  (typeof globalThis)['CustomElementRegistry'];
+
+// Ponyfill for PromiseWithResolvers, remove once we can assume its presence.
+type PromiseWithResolvers<T> = {
+  promise: Promise<T>;
+  resolve: (value: T) => void;
+  reject: (reason?: unknown) => void;
+};
+function promiseWithResolvers<T>(): PromiseWithResolvers<T> {
+  let resolve: (value: T) => void;
+  let reject: (reason?: unknown) => void;
+  const promise = new Promise<T>((res, rej) => {
+    resolve = res;
+    reject = rej;
+  });
+  return {promise, resolve: resolve!, reject: reject!};
```

Comment:
fyi, you can do slightly more compact non-null assertions:

```suggestion
  let resolve!: (value: T) => void;
  let rejec!t: (reason?: unknown) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return {promise, resolve, reject};
```